src/check/parse/parser.zig

```diff
@@ -0,0 +1,490 @@
+const std = @import("std");
+
+const IR = @import("ir.zig");
+const NodeList = IR.NodeList;
+
+const tokenize = @import("tokenize.zig");
+const TokenIndex = tokenize.TokenIndex;
+const TokenizedBuffer = tokenize.TokenizedBuffer;
+const Token = tokenize.Token;
+
+pub const Parser = @This();
+
+allocator: std.mem.Allocator,
+pos: TokenIndex,
+tokens: TokenizedBuffer,
+store: IR.NodeStore,
+scratch: std.ArrayList(IR.Node.Index),
+diagnostics: std.ArrayList(IR.Diagnostic),
+
+pub fn init(gpa: std.mem.Allocator, tokens: TokenizedBuffer) !Parser {
+    const estimated_node_count = (tokens.tokens.len + 2) / 2;
+    const store = try IR.NodeStore.initWithCapacity(gpa, estimated_node_count);
+
+    return Parser{
+        .allocator = gpa,
+        .pos = 0,
+        .tokens = tokens,
+        .store = store,
+        .scratch = std.ArrayList(IR.Node.Index).init(gpa),
+        .diagnostics = std.ArrayList(IR.Diagnostic).init(gpa),
+    };
+}
+
+pub fn deinit(parser: *Parser) void {
+    parser.scratch.deinit();
+    parser.diagnostics.deinit();
+}
+
+fn test_parser(source: []const u8, run: fn (parser: Parser) std.mem.AllocatorError!void) !void {
+    const gc = try @import("GenCatData").GenCatData.init(std.testing.allocator);
+    defer gc.deinit();
+    const messages = [128]tokenize.Diagnostic;
+    const tokenizer = try tokenize.Tokenizer.init(source, messages[0..], gc, std.testing.allocator);
+    try tokenizer.tokenize();
+    const tok_result = try tokenizer.finalize_and_deinit();
+    defer tok_result.tokens.deinit();
+    const parser = Parser{
+        .allocator = std.testing.allocator,
+        .pos = 0,
+        .tokens = tok_result.tokens,
+        .store = IR.NodeStore.initWithCapacity(std.testing.allocator, source.len),
+        .scratch = std.ArrayList(IR.Node.Index).init(std.testing.allocator),
+        .diagnostics = std.ArrayList(IR.Diagnostic).init(std.testing.allocator),
+    };
+    defer parser.store.deinit();
+    defer parser.scratch.deinit();
+    defer parser.diagnostics.deinit();
+    try run(parser);
+}
+
+pub fn advance(self: *Parser) void {
+    if (self.pos >= self.tokens.tokens.len) {
+        return;
+    }
+    self.pos += 1;
+}
+
+pub fn peek(self: *Parser) Token.Tag {
+    if (self.pos >= self.tokens.tokens.len) {
+        return .EndOfFile;
+    }
+    return self.tokens.tokens.items(.tag)[self.pos];
+}
+
+pub fn peekNext(self: Parser) Token.Tag {
+    const next = self.pos + 1;
+    if (next >= self.tokens.tokens.len) {
+        return .EndOfFile;
+    }
+    return self.tokens.tokens.items(.tag)[next];
+}
+
+// If the next token is a newline, consume it
+// Returns the indent level of the next line if it is a newline, otherwise null
+pub fn consumeNewline(self: *Parser) ?u16 {
+    if (self.peek() != .Newline) {
+        return null;
+    }
```

Comment:
From a parser perspective, I think comments and newlines should be treated identically, no? Why only look for newlines here?

Reply:
There is no comment node